lib/auth/grpcserver.go

```diff
@@ -1886,6 +1886,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
 	return resp, nil
 }
 
+// GetToken retrieves a token by name.
+func (g *GRPCServer) GetToken(ctx context.Context, req *types.ResourceRequest) (*types.ProvisionTokenV2, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	token, err := auth.ServerWithRoles.GetToken(ctx, req.Name)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	provisionTokenV2, ok := token.(*types.ProvisionTokenV2)
+	if !ok {
+		return nil, trail.ToGRPC(trace.Errorf("encountered unexpected token type: %T", token))
+	}
+	return provisionTokenV2, nil
+}
+
+// GetTokens retrieves all tokens.
+func (g *GRPCServer) GetTokens(ctx context.Context, _ *empty.Empty) (*types.ProvisionTokenV2List, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	tokens, err := auth.ServerWithRoles.GetTokens(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	provisionTokensV2 := make([]*types.ProvisionTokenV2, len(tokens))
+	for i, t := range tokens {
+		var ok bool
+		if provisionTokensV2[i], ok = t.(*types.ProvisionTokenV2); !ok {
+			return nil, trail.ToGRPC(trace.BadParameter("unexpected type %T", t))
+		}
+	}
+	return &types.ProvisionTokenV2List{
+		ProvisionTokens: provisionTokensV2,
+	}, nil
+}
+
+// UpsertToken upserts a token.
+func (g *GRPCServer) UpsertToken(ctx context.Context, token *types.ProvisionTokenV2) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	if err = auth.ServerWithRoles.UpsertToken(ctx, token); err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	g.Debugf("token upserted")
```

Comment:
This is not logged by most other APIs - is this required? Same for `DeleteToken`